File: patterns/_tests/CreationalPatterns/AbstractFactory/typescript/integration.test.ts
// ============================================================
//  Integration Tests — Abstract Factory: Biome World Builder
//
//  What we test here:
//    · Real factory singletons produce correct, consistent output
//    · WorldGenerator generates the full product pipeline end-to-end
//    · Swapping the factory changes every product (terrain + enemies + weather + loot)
//    · Product families stay isolated (no cross-biome contamination)
//    · Prototype registration at runtime propagates into the next generation
//
//  No stubs. All concrete classes interact exactly as in production.
//
//  Run: npx mocha --require ts-node/register integration.test.ts
// ============================================================

// SECTION:: Imports
import { describe, it, before } from 'mocha';
import assert from 'assert';
import { captureOutput, findMissing } from '../../../_helpers/typescript/helpers';
import {
  DesertBiomeKit,
  ArcticBiomeKit,
  ForestBiomeKit,
  WorldGenerator,
} from '../../../../CreationalPatterns/AbstractFactory/typescript/typescript';

// SECTION:: WorldGenerator — Desert Integration
describe('WorldGenerator — Desert Biome', () => {
  let lines: string[];

  before(() => {
    lines = captureOutput(() => new WorldGenerator(DesertBiomeKit.getInstance()).generate());
  });

  it('output contains desert terrain name', () => {
    assert.ok(lines.some(l => l.includes('Sand Dunes')));
  });

  it('output contains all 3 default desert enemies', () => {
    assert.strictEqual(findMissing(lines, ['Scorpion', 'Sand Worm', 'Desert Bandit']), null);
  });

  it('output contains Sandstorm weather', () => {
    assert.ok(lines.some(l => l.includes('Sandstorm')));
  });

  it('output contains desert loot table name', () => {
    assert.ok(lines.some(l => l.includes('Desert Loot Table')));
  });
});

// SECTION:: WorldGenerator — Arctic & Forest Integration
describe('WorldGenerator — Arctic Biome', () => {
  let lines: string[];

  before(() => {
    lines = captureOutput(() => new WorldGenerator(ArcticBiomeKit.getInstance()).generate());
  });

  it('output contains Frozen Tundra terrain', () => {
    assert.ok(lines.some(l => l.includes('Frozen Tundra')));
  });

  it('output contains all 3 arctic enemies', () => {
    assert.strictEqual(findMissing(lines, ['Polar Bear', 'Ice Wolf', 'Frost Troll']), null);
  });

  it('output contains Blizzard weather', () => {
    assert.ok(lines.some(l => l.includes('Blizzard')));
  });
});

describe('WorldGenerator — Forest Biome', () => {
  let lines: string[];

  before(() => {
    lines = captureOutput(() => new WorldGenerator(ForestBiomeKit.getInstance()).generate());
  });

  it('output contains Dense Forest Floor terrain', () => {
    assert.ok(lines.some(l => l.includes('Dense Forest Floor')));
  });

  it('output contains all 3 forest enemies', () => {
    assert.strictEqual(findMissing(lines, ['Wolf', 'Wild Boar', 'Forest Bandit']), null);
  });
});

// SECTION:: Product Family Consistency
describe('Product Family Consistency', () => {
  it('swapping factory changes every part of the output', () => {
    const desert = captureOutput(() => new WorldGenerator(DesertBiomeKit.getInstance()).generate()).join('\n');
    const arctic  = captureOutput(() => new WorldGenerator(ArcticBiomeKit.getInstance()).generate()).join('\n');

    assert.ok(desert.includes('Sand Dunes'));
    assert.ok(!desert.includes('Frozen Tundra'));
    assert.ok(arctic.includes('Frozen Tundra'));
    assert.ok(!arctic.includes('Sand Dunes'));
    assert.ok(!desert.includes('Polar Bear'));
    assert.ok(!arctic.includes('Scorpion'));
  });

  it('desert loot table never appears in arctic output', () => {
    const arctic = captureOutput(() => new WorldGenerator(ArcticBiomeKit.getInstance()).generate()).join('\n');
    assert.ok(!arctic.includes('Desert Loot Table'));
    assert.ok(arctic.includes('Arctic Loot Table'));
  });

  it('weather events are biome-specific', () => {
    const desert = captureOutput(() => new WorldGenerator(DesertBiomeKit.getInstance()).generate()).join('\n');
    const arctic  = captureOutput(() => new WorldGenerator(ArcticBiomeKit.getInstance()).generate()).join('\n');
    assert.ok(desert.includes('heat_wave'));
    assert.ok(arctic.includes('whiteout'));
    assert.ok(!desert.includes('whiteout'));
    assert.ok(!arctic.includes('heat_wave'));
  });
});

// SECTION:: Prototype Registration
describe('Prototype Registration (runtime enemy injection)', () => {
  it('registerEnemy adds a new enemy to the next generate() call', () => {
    const kit    = DesertBiomeKit.getInstance();
    const before = captureOutput(() => new WorldGenerator(kit).generate()).join('\n');

    kit.registerEnemy({
      name: 'Giant Scorpion (Boss)', health: 400, damage: 60,
      attack: 'Tail sweep — hits all nearby players', drop: 'Ancient Venom Sac',
      clone() { return this; },
    });

    const after = captureOutput(() => new WorldGenerator(kit).generate()).join('\n');
    assert.ok(!before.includes('Giant Scorpion'));
    assert.ok(after.includes('Giant Scorpion'));
  });

  it('registration is isolated — arctic kit is unaffected by desert registration', () => {
    const arctic = captureOutput(() => new WorldGenerator(ArcticBiomeKit.getInstance()).generate()).join('\n');
    assert.ok(!arctic.includes('Giant Scorpion'));
  });

  it('registered enemy is cloned on generate() — clone() is called', () => {
    let cloneCalled = false;
    ForestBiomeKit.getInstance().registerEnemy({
      name: 'Tracker', health: 1, damage: 1, attack: 'Track', drop: 'Data',
      clone() { cloneCalled = true; return this; },
    });
    captureOutput(() => new WorldGenerator(ForestBiomeKit.getInstance()).generate());
    assert.ok(cloneCalled);
  });
});

File: patterns/_tests/CreationalPatterns/AbstractFactory/ruby/integration.test.rb
# ============================================================
#  Integration Tests — Abstract Factory: Biome World Builder
#
#  What we test here:
#    · Real factory singletons produce correct, consistent output
#    · WorldGenerator generates the full product pipeline end-to-end
#    · Swapping the factory changes every product
#    · Product families stay isolated (no cross-biome contamination)
#    · Prototype registration at runtime propagates into the next generation
#
#  No stubs. All concrete classes interact exactly as in production.
#
#  Run: rspec integration_spec.rb
# ============================================================

# SECTION:: Imports
require 'stringio'
require_relative '../../../../CreationalPatterns/AbstractFactory/ruby/ruby'

def capture_generate(kit)
  old = $stdout
  $stdout = StringIO.new
  WorldGenerator.new(kit).generate
  $stdout.string
ensure
  $stdout = old
end

# SECTION:: WorldGenerator — Desert Integration
RSpec.describe 'WorldGenerator — Desert Biome' do
  let(:output) { capture_generate(DesertBiomeKit.instance) }

  it 'contains desert terrain name'       do expect(output).to include('Sand Dunes')       end
  it 'contains Scorpion enemy'            do expect(output).to include('Scorpion')          end
  it 'contains Sand Worm enemy'           do expect(output).to include('Sand Worm')         end
  it 'contains Desert Bandit enemy'       do expect(output).to include('Desert Bandit')     end
  it 'contains Sandstorm weather'         do expect(output).to include('Sandstorm')         end
  it 'contains Desert Loot Table'         do expect(output).to include('Desert Loot Table') end
end

# SECTION:: WorldGenerator — Arctic & Forest Integration
RSpec.describe 'WorldGenerator — Arctic Biome' do
  let(:output) { capture_generate(ArcticBiomeKit.instance) }

  it 'contains Frozen Tundra terrain'     do expect(output).to include('Frozen Tundra')    end
  it 'contains Polar Bear enemy'          do expect(output).to include('Polar Bear')        end
  it 'contains Ice Wolf enemy'            do expect(output).to include('Ice Wolf')          end
  it 'contains Frost Troll enemy'         do expect(output).to include('Frost Troll')       end
  it 'contains Blizzard weather'          do expect(output).to include('Blizzard')          end
end

RSpec.describe 'WorldGenerator — Forest Biome' do
  let(:output) { capture_generate(ForestBiomeKit.instance) }

  it 'contains Dense Forest Floor terrain' do expect(output).to include('Dense Forest Floor') end
  it 'contains Wolf enemy'                 do expect(output).to include('Wolf')               end
  it 'contains Wild Boar enemy'            do expect(output).to include('Wild Boar')          end
  it 'contains Forest Bandit enemy'        do expect(output).to include('Forest Bandit')      end
end

# SECTION:: Product Family Consistency
RSpec.describe 'Product Family Consistency' do
  it 'swapping factory changes every part of the output' do
    desert = capture_generate(DesertBiomeKit.instance)
    arctic  = capture_generate(ArcticBiomeKit.instance)

    expect(desert).to     include('Sand Dunes')
    expect(desert).not_to include('Frozen Tundra')
    expect(arctic).to     include('Frozen Tundra')
    expect(arctic).not_to include('Sand Dunes')
    expect(desert).not_to include('Polar Bear')
    expect(arctic).not_to include('Scorpion')
  end

  it 'desert loot table never appears in arctic output' do
    arctic = capture_generate(ArcticBiomeKit.instance)
    expect(arctic).not_to include('Desert Loot Table')
    expect(arctic).to     include('Arctic Loot Table')
  end

  it 'weather events are biome-specific' do
    desert = capture_generate(DesertBiomeKit.instance)
    arctic  = capture_generate(ArcticBiomeKit.instance)
    expect(desert).to     include('heat_wave')
    expect(arctic).to     include('whiteout')
    expect(desert).not_to include('whiteout')
    expect(arctic).not_to include('heat_wave')
  end
end

# SECTION:: Prototype Registration
RSpec.describe 'Prototype Registration (runtime enemy injection)' do
  it 'registered enemy appears in the next generate call' do
    kit    = DesertBiomeKit.instance
    before = capture_generate(kit)

    giant = Struct.new(:name, :health, :damage, :attack, :drop) { def clone = dup }
    kit.register_enemy(giant.new('Giant Scorpion (Boss)', 400, 60,
                                 'Tail sweep', 'Ancient Venom Sac'))

    after = capture_generate(kit)
    expect(before).not_to include('Giant Scorpion')
    expect(after).to      include('Giant Scorpion')
  end

  it 'registration is isolated — arctic kit unaffected' do
    arctic = capture_generate(ArcticBiomeKit.instance)
    expect(arctic).not_to include('Giant Scorpion')
  end

  it 'registered enemy is cloned on each generate call' do
    clone_calls = 0
    tracker = Object.new
    tracker.define_singleton_method(:name)   { 'Tracker' }
    tracker.define_singleton_method(:health) { 1 }
    tracker.define_singleton_method(:damage) { 1 }
    tracker.define_singleton_method(:attack) { 'Track' }
    tracker.define_singleton_method(:drop)   { 'Data' }
    tracker.define_singleton_method(:clone)  { clone_calls += 1; tracker }

    ForestBiomeKit.instance.register_enemy(tracker)
    capture_generate(ForestBiomeKit.instance)
    expect(clone_calls).to be >= 1
  end
end
